examples/CustomContainer.tsx

```diff
@@ -0,0 +1,86 @@
+import * as React from 'react';
+import { List, arrayMove } from '../src/index';
+
+const Basic: React.FC<{
+  container: Element;
+}> = (props) => {
+  const { container } = props;
+  const [items, setItems] = React.useState([
+    'Item 1',
+    'Item 2',
+    'Item 3',
+    'Item 4',
+    'Item 5',
+    'Item 6'
+  ]);
+  console.log(container);
+
+  return (
+    <div
+      style={{
+        maxWidth: '300px',
+        margin: '0px auto',
+        backgroundColor: '#F7F7F7',
+        padding: '3em'
+      }}
+    >
+      <div>
+        <List
+          values={items}
+          onChange={({ oldIndex, newIndex }) =>
+            setItems(arrayMove(items, oldIndex, newIndex))
+          }
+          renderList={({ children, props, isDragged }) => (
+            <ul
+              {...props}
+              style={{
+                padding: 0,
+                cursor: isDragged ? 'grabbing' : undefined
+              }}
+            >
+              {children}
+            </ul>
+          )}
+          renderItem={({ value, props, isDragged, isSelected }) => (
+            <li
+              {...props}
+              style={{
+                ...props.style,
+                padding: '1.5em',
+                margin: '0.5em 0em',
+                listStyleType: 'none',
+                cursor: isDragged ? 'grabbing' : 'grab',
+                border: '2px solid #CCC',
+                boxShadow: '3px 3px #AAA',
+                color: '#333',
+                borderRadius: '5px',
+                fontFamily: 'Arial, "Helvetica Neue", Helvetica, sans-serif',
+                backgroundColor: isDragged || isSelected ? '#EEE' : '#FFF'
+              }}
+            >
+              {value}
+            </li>
+          )}
+          container={container}
+        />
+      </div>
+    </div>
+  );
+};
+
+const CustomContainer: React.FC = () => {
```

Comment:
Can you just merge both components in this file? I don't see why it needs to be split into two and makes it less readable. You also don't need `useEffect` to set a ref?

Reply:
Thank you for review.

Sure merged, `useEffect` is used because:
> Mutating the .current property doesn’t cause a re-render.

Reply:
There is no need to re-render. This works just fine:

```jsx
import * as React from 'react';
import { List, arrayMove } from '../src/index';

const CustomContainer: React.FC = () => {
  const wrapper = React.useRef<HTMLDivElement>(null);
  const [items, setItems] = React.useState([
    'Item 1',
    'Item 2',
    'Item 3',
    'Item 4',
    'Item 5',
    'Item 6'
  ]);

  return (
    <div className="wrapper" ref={wrapper}>
      <div
        style={{
          maxWidth: '300px',
          margin: '0px auto',
          backgroundColor: '#F7F7F7',
          padding: '3em'
        }}
      >
        <div>
          <List
            values={items}
            onChange={({ oldIndex, newIndex }) =>
              setItems(arrayMove(items, oldIndex, newIndex))
            }
            renderList={({ children, props, isDragged }) => (
              <ul
                {...props}
                style={{
                  padding: 0,
                  cursor: isDragged ? 'grabbing' : undefined
                }}
              >
                {children}
              </ul>
            )}
            renderItem={({ value, props, isDragged, isSelected }) => (
              <li
                {...props}
                style={{
                  ...props.style,
                  padding: '1.5em',
                  margin: '0.5em 0em',
                  listStyleType: 'none',
                  cursor: isDragged ? 'grabbing' : 'grab',
                  border: '2px solid #CCC',
                  boxShadow: '3px 3px #AAA',
                  color: '#333',
                  borderRadius: '5px',
                  fontFamily: 'Arial, "Helvetica Neue", Helvetica, sans-serif',
                  backgroundColor: isDragged || isSelected ? '#EEE' : '#FFF'
                }}
              >
                {value}
              </li>
            )}
            container={wrapper.current}
          />
        </div>
      </div>
    </div>
  );
};

export default CustomContainer;
```

Reply:
In this approach first render will render li in body.
When component will update it will be rendered in wrapper.

![Peek 2020-04-30 11-04](https://user-images.githubusercontent.com/8374447/80698692-c1f44d80-8ad2-11ea-8e48-1146eaa274f1.gif)

Reply:
Ah, you are right. Ok, still we can clean it up a bit:
- removing unnecessary divs
- useEffect second param should be `[wrapper.current]` since that's the only dependency inside of the effect

```jsx
import * as React from 'react';
import { List, arrayMove } from '../src/index';

const CustomContainer: React.FC = () => {
  const wrapper = React.useRef<HTMLDivElement>(null);
  const [items, setItems] = React.useState([
    'Item 1',
    'Item 2',
    'Item 3',
    'Item 4',
    'Item 5',
    'Item 6'
  ]);
  const [container, setContainer] = React.useState<Element | null>(null);
  React.useEffect(() => {
    setContainer(wrapper.current);
  }, [wrapper.current]);

  return (
    <div
      ref={wrapper}
      style={{
        maxWidth: '300px',
        margin: '0px auto',
        backgroundColor: '#F7F7F7',
        padding: '3em'
      }}
    >
      <List
        container={container}
        values={items}
        onChange={({ oldIndex, newIndex }) =>
          setItems(arrayMove(items, oldIndex, newIndex))
        }
        renderList={({ children, props, isDragged }) => (
          <ul
            {...props}
            style={{
              padding: 0,
              cursor: isDragged ? 'grabbing' : undefined
            }}
          >
            {children}
          </ul>
        )}
        renderItem={({ value, props, isDragged, isSelected }) => (
          <li
            {...props}
            style={{
              ...props.style,
              padding: '1.5em',
              margin: '0.5em 0em',
              listStyleType: 'none',
              cursor: isDragged ? 'grabbing' : 'grab',
              border: '2px solid #CCC',
              boxShadow: '3px 3px #AAA',
              color: '#333',
              borderRadius: '5px',
              fontFamily: 'Arial, "Helvetica Neue", Helvetica, sans-serif',
              backgroundColor: isDragged || isSelected ? '#EEE' : '#FFF'
            }}
          >
            {value}
          </li>
        )}
      />
    </div>
  );
};

export default CustomContainer;
```

Reply:
Cleaned :)

Reply:
Thanks!